### PR TITLE
feat: add stream methods for `Page`

### LINKED ISSUE #1426: `Page` do not easily support Java Streams
Add stream support for `com.google.api.gax.paging.Page`

### PR DESCRIPTION
**Pull request by JoeWang1127**
Fixes #1426  ☕️

### CODE REVIEW

**Comment by JoeWang1127** [comment]
Convert to draft for adding integration tests.

**Comment by suztomo** [comment]
Don't merge this yet. We want to control when to introduce this new feature.

File: gax-java/gax/clirr-ignored-differences.xml
Comment (suztomo) [inline gax-java/gax/clirr-ignored-differences.xml:5]: Add a comment why this is ok.

Comment (JoeWang1127) [reply gax-java/gax/clirr-ignored-differences.xml:6]: done

**Comment by JoeWang1127** [comment]
I'll merge main branch into this branch when it's ready to merge.

File: gax-java/gax/clirr-ignored-differences.xml
Comment (suztomo) [inline gax-java/gax/clirr-ignored-differences.xml:8]: It's strange that CLIRR is complaining the change. Adding default method in interface shouldn't cause a breaking changes.

Comment (suztomo) [reply gax-java/gax/clirr-ignored-differences.xml:8]: @JoeWang1127 Just in case, would you build this gax-java locally (with -SNAPSHOT version) and try to use in java-storage to see anything breaks? You ran some storage samples that you used to understand the request.

Comment (JoeWang1127) [reply gax-java/gax/clirr-ignored-differences.xml:8]: In [Specifying ignored differences](https://www.mojohaus.org/clirr-maven-plugin/examples/ignored-differences.html) of Clirr Maven Plugin doc, 7012 has the following definition:
```
7012 (Method Added to Interface): className, method
```
So adding a method to interface, default or not, is a breaking change to this rule.

Also, there's 7013:
```
7013 (Abstract Method Added to Class): className, method
```
which is not complain.

**Comment by JoeWang1127** [comment]
@suztomo I've verified the change in the demo project (on branch [`use-stream`](https://github.com/JoeWang1127/storage-demo/tree/use-stream)).

**Comment by suztomo** [comment]
Thank you.

File: gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java
Comment (suztomo) [inline gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java:73]: @BenWhitehead Would you elaborate what's not ideal? Parallel Stream.

File: gax-java/gax/clirr-ignored-differences.xml
Comment (BenWhitehead) [reply gax-java/gax/clirr-ignored-differences.xml:8]: Clirr isn't aware of java 8 default methods. It is looking at the bytecode from the perspective of java7 binary compatibility. When adding a default method it is safe to specify an ignore rule.

File: gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java
Comment (BenWhitehead) [reply gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java:73]: I think this is great for a first step of getting it on the api and saving our customers from having to do the same hoop jumping.

In the case of future improvements, right now this approach will not allow of parallel processing the same way a natively parallel friendly stream would. In all actuality, Page isn't fully parallelizable due to the fact that each page contains the `nextPageToken` necessary in order to fetch the next page. But, Page could be more parallel friendly where processing of the actual resources can happen on a thread separate from the thread performing the actual get.

Comment (suztomo) [reply gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java:73]: @JoeWang1127 Do you have question or comment about parallelism? (I don't think of an implementation as of now)

Comment (BenWhitehead) [reply gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java:73]: I agree, parallelism does not need to block this PR and can be an additional improvement in the future.

Comment (JoeWang1127) [reply gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java:73]: @suztomo I'm not understand about the separate thread in Ben's comment:
```
Page could be more parallel friendly where processing of the actual resources can happen on a thread separate from the thread performing the actual get.
```

Comment (suztomo) [reply gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java:73]: This https://docs.oracle.com/javase/tutorial/collections/streams/parallelism.html#executing_streams_in_parallel

File: gax-java/gax/clirr-ignored-differences.xml
Comment (JoeWang1127) [reply gax-java/gax/clirr-ignored-differences.xml:8]: @BenWhitehead Thanks for the clarification.

File: gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java
Comment (JoeWang1127) [reply gax-java/gax/src/main/java/com/google/api/gax/paging/Page.java:73]: Thanks for the reference.

I don't have a question about it.

File: gax-java/gax/src/test/java/com/google/api/gax/rpc/PagingTest.java
Comment (blakeli0) [inline gax-java/gax/src/test/java/com/google/api/gax/rpc/PagingTest.java:137]: Can we update the test names based on the [best practices](https://engdoc.corp.google.com/eng/doc/devguide/testing/unit/best-practices.md?cl=head&polyglot=java#naming) and a recent [TotT](https://engdoc.corp.google.com/eng/doc/tott/episodes/660.md?cl=head)?

Comment (JoeWang1127) [reply gax-java/gax/src/test/java/com/google/api/gax/rpc/PagingTest.java]: done

**Review by blakeli0 — APPROVED** [review]

**Pull request merged**
